File: Algorithms/dynamic program/Stair.java
public class Stair {
    static int f[] = new int[201];
    static int g[] = new int[201];

    private static void stair(int n) {
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= n; j++) {
                if (i < 2) {
                    f[j] = 0;
                } else if (j > i) {
                    int x = g[j];
                    int y = g[j - i];
                    if ((j - i <= i * (i - 1) / 2) && (j - i < i)) {
                        y += 1;
                    }
                    f[j] = x + y;
                } else {
                    f[j] = g[j];
                }
            }
            g = f.clone();
        }
    }

    public static void main(String[] args) {
        int n = 200;
        stair(n);
        System.out.println(f[n]);
    }
}

File: algorithms/dynamic_program/stair.cpp
#include<iostream>

using namespace std;
	
	int stair[100001]={0};
	long long res[100001]={0};
	

int main(){
	//initial
	int stairNumber; cin>>stairNumber;
	int brokenStairNumber; cin>>brokenStairNumber;
	for(int i=0; i<brokenStairNumber; i++){
		int pos; cin>>pos;
		stair[pos]=1;
	}
	//Dynamic planning
	res[0]=0;
	res[1]=1;
	for(int i=2; i<=stairNumber; i++){
		if(stair[i]==0){
			res[i]= res[i-1]+ res[i-2];
			res[i]=res[i]%14062008;
		}
	}
	//output
	cout<<res[stairNumber];
	
	return 0;
}
